File: apps/mobile/client-apple/VettaKit/Sources/VettaKit/Protocol/RemoteCrypto.swift
import CryptoKit
import Foundation
import Security

/// End-to-end encryption for protocol v2, byte-compatible with `crypto.ts`:
/// X25519 identity + ephemeral keys, HKDF-SHA256 and XChaCha20-Poly1305.
/// CryptoKit supplies X25519, HKDF and the IETF ChaCha20-Poly1305 AEAD; the
/// extended 24-byte nonce is the standard HChaCha20 subkey construction.
public struct RemoteIdentityKeyPair: Sendable {
	public let secretKey: Data
	public let publicKey: Data

	public init(secretKey: Data) throws {
		guard secretKey.count == 32 else { throw RemoteProtocolError("identity secret must be 32 bytes") }
		let privateKey = try Curve25519.KeyAgreement.PrivateKey(rawRepresentation: secretKey)
		self.secretKey = secretKey
		publicKey = privateKey.publicKey.rawRepresentation
	}

	public static func generate(randomBytes: RandomBytes = RemoteCrypto.randomBytes) -> RemoteIdentityKeyPair {
		// A 32-byte random scalar is always a valid X25519 secret (clamping happens inside).
		try! RemoteIdentityKeyPair(secretKey: randomBytes(32))
	}
}

public typealias RandomBytes = (Int) -> Data

public struct RemoteSessionKeys: Sendable {
	public let sendKey: Data
	public let receiveKey: Data
}

public enum RemoteCrypto {
	static let nonceLength = 24
	static let keyLength = 32
	static let infoPrefix = "vetta-remote-v2"
	public static let sealedAssociatedData = "vetta-remote-v\(remoteProtocolVersion)"

	nonisolated public static func randomBytes(_ count: Int) -> Data {
		var bytes = [UInt8](repeating: 0, count: count)
		let status = SecRandomCopyBytes(kSecRandomDefault, count, &bytes)
		precondition(status == errSecSuccess, "system random generator failed")
		return Data(bytes)
	}

	static func sharedSecret(_ secretKey: Data, _ peerPublicKey: Data) throws -> Data {
		let privateKey = try Curve25519.KeyAgreement.PrivateKey(rawRepresentation: secretKey)
		let peer = try Curve25519.KeyAgreement.PublicKey(rawRepresentation: peerPublicKey)
		let shared = try privateKey.sharedSecretFromKeyAgreement(with: peer)
		return shared.withUnsafeBytes { Data($0) }
	}

	static func hkdf(_ ikm: Data, salt: Data, info: String, length: Int) -> Data {
		let key = HKDF<SHA256>.deriveKey(
			inputKeyMaterial: SymmetricKey(data: ikm),
			salt: salt,
			info: Data(info.utf8),
			outputByteCount: length
		)
		return key.withUnsafeBytes { Data($0) }
	}

	public static func deriveSessionKeys(
		role: RemoteRole,
		identity: RemoteIdentityKeyPair,
		ephemeral: RemoteIdentityKeyPair,
		peerIdentityKey: Data,
		peerEphemeralKey: Data
	) throws -> RemoteSessionKeys {
		let ephemeralShared = try sharedSecret(ephemeral.secretKey, peerEphemeralKey)
		let staticToPeerEphemeral = try sharedSecret(identity.secretKey, peerEphemeralKey)
		let ephemeralToPeerStatic = try sharedSecret(ephemeral.secretKey, peerIdentityKey)
		// Order the static/ephemeral mixes by role so both sides compute the same input.
		let mobileFirst = role == .mobile
		let ikm = ephemeralShared
			+ (mobileFirst ? staticToPeerEphemeral : ephemeralToPeerStatic)
			+ (mobileFirst ? ephemeralToPeerStatic : staticToPeerEphemeral)
		let salt = mobileFirst
			? identity.publicKey + peerIdentityKey + ephemeral.publicKey + peerEphemeralKey
			: peerIdentityKey + identity.publicKey + peerEphemeralKey + ephemeral.publicKey
		let material = hkdf(ikm, salt: salt, info: "\(infoPrefix)/session", length: keyLength * 2)
		let mobileToDesktop = material.prefix(keyLength)
		let desktopToMobile = material.suffix(keyLength)
		return mobileFirst
			? RemoteSessionKeys(sendKey: Data(mobileToDesktop), receiveKey: Data(desktopToMobile))
			: RemoteSessionKeys(sendKey: Data(desktopToMobile), receiveKey: Data(mobileToDesktop))
	}

	// MARK: XChaCha20-Poly1305

	public static func xchachaSeal(key: Data, nonce: Data, plaintext: Data, associatedData: Data) throws -> Data {
		let (subkey, ietfNonce) = try extendedNonce(key: key, nonce: nonce)
		let box = try ChaChaPoly.seal(plaintext, using: SymmetricKey(data: subkey), nonce: ietfNonce, authenticating: associatedData)
		return box.ciphertext + box.tag
	}

	public static func xchachaOpen(key: Data, nonce: Data, ciphertext: Data, associatedData: Data) throws -> Data {
		guard ciphertext.count >= 16 else { throw RemoteProtocolError("ciphertext too short") }
		let (subkey, ietfNonce) = try extendedNonce(key: key, nonce: nonce)
		let box = try ChaChaPoly.SealedBox(
			nonce: ietfNonce,
			ciphertext: ciphertext.prefix(ciphertext.count - 16),
			tag: ciphertext.suffix(16)
		)
		return try ChaChaPoly.open(box, using: SymmetricKey(data: subkey), authenticating: associatedData)
	}

	private static func extendedNonce(key: Data, nonce: Data) throws -> (Data, ChaChaPoly.Nonce) {
		guard key.count == 32, nonce.count == nonceLength else { throw RemoteProtocolError("invalid key or nonce length") }
		let subkey = hchacha20(key: [UInt8](key), nonce: [UInt8](nonce.prefix(16)))
		let ietf = try ChaChaPoly.Nonce(data: Data([0, 0, 0, 0]) + nonce.suffix(8))
		return (Data(subkey), ietf)
	}

	static func hchacha20(key: [UInt8], nonce: [UInt8]) -> [UInt8] {
		func word(_ bytes: [UInt8], _ offset: Int) -> UInt32 {
			UInt32(bytes[offset]) | UInt32(bytes[offset + 1]) << 8 | UInt32(bytes[offset + 2]) << 16 | UInt32(bytes[offset + 3]) << 24
		}
		var state: [UInt32] = [0x6170_7865, 0x3320_646E, 0x7962_2D32, 0x6B20_6574]
		for index in 0 ..< 8 { state.append(word(key, index * 4)) }
		for index in 0 ..< 4 { state.append(word(nonce, index * 4)) }
		func quarter(_ a: Int, _ b: Int, _ c: Int, _ d: Int) {
			state[a] &+= state[b]; state[d] ^= state[a]; state[d] = state[d].rotateLeft(16)
			state[c] &+= state[d]; state[b] ^= state[c]; state[b] = state[b].rotateLeft(12)
			state[a] &+= state[b]; state[d] ^= state[a]; state[d] = state[d].rotateLeft(8)
			state[c] &+= state[d]; state[b] ^= state[c]; state[b] = state[b].rotateLeft(7)
		}
		for _ in 0 ..< 10 {
			quarter(0, 4, 8, 12); quarter(1, 5, 9, 13); quarter(2, 6, 10, 14); quarter(3, 7, 11, 15)
			quarter(0, 5, 10, 15); quarter(1, 6, 11, 12); quarter(2, 7, 8, 13); quarter(3, 4, 9, 14)
		}
		var output: [UInt8] = []
		for index in [0, 1, 2, 3, 12, 13, 14, 15] {
			let value = state[index]
			output += [UInt8(value & 0xFF), UInt8(value >> 8 & 0xFF), UInt8(value >> 16 & 0xFF), UInt8(value >> 24 & 0xFF)]
		}
		return output
	}

	// MARK: Sealed frames

	public static func sealFrame(key: Data, frame: RemoteFrame, associatedData: String = sealedAssociatedData, randomBytes: RandomBytes = RemoteCrypto.randomBytes) throws -> RemoteSealed {
		let nonce = randomBytes(nonceLength)
		let ciphertext = try xchachaSeal(
			key: key,
			nonce: nonce,
			plaintext: Data(frame.json.serialized().utf8),
			associatedData: Data(associatedData.utf8)
		)
		return RemoteSealed(nonce: Base64URL.encode(nonce), ciphertext: Base64URL.encode(ciphertext))
	}

	public static func openFrame(key: Data, sealed: RemoteSealed, associatedData: String = sealedAssociatedData) throws -> RemoteFrame {
		let plaintext: Data
		do {
			plaintext = try xchachaOpen(
				key: key,
				nonce: try Base64URL.decode(sealed.nonce),
				ciphertext: try Base64URL.decode(sealed.ciphertext),
				associatedData: Data(associatedData.utf8)
			)
		} catch {
			throw RemoteProtocolError("sealed frame failed authentication")
		}
		let value: JSONValue
		do {
			value = try JSONValue.parse(plaintext)
		} catch {
			throw RemoteProtocolError("sealed frame is not valid JSON")
		}
		return try RemoteFrame.decodeSession(value)
	}

	/// Six digits both ends display during a manual pairing; derived from both identity keys.
	public static func verificationCode(_ identityKeyA: Data, _ identityKeyB: Data) -> String {
		let ordered = compare(identityKeyA, identityKeyB) <= 0 ? identityKeyA + identityKeyB : identityKeyB + identityKeyA
		let digest = [UInt8](hkdf(ordered, salt: Data(), info: "\(infoPrefix)/sas", length: 4))
		let value = UInt32(digest[0]) << 24 | UInt32(digest[1]) << 16 | UInt32(digest[2]) << 8 | UInt32(digest[3])
		let code = String(value % 1_000_000)
		return String(repeating: "0", count: max(0, 6 - code.count)) + code
	}

	public static func sha256Hex(_ text: String) -> String {
		SHA256.hash(data: Data(text.utf8)).map { String(format: "%02x", $0) }.joined()
	}

	public static func randomToken(bytes: Int = 32, randomBytes: RandomBytes = RemoteCrypto.randomBytes) -> String {
		Base64URL.encode(randomBytes(bytes))
	}

	public static func decodePublicKey(_ text: String, field: String = "public key") throws -> Data {
		let bytes = try Base64URL.decode(text)
		guard bytes.count == keyLength else { throw RemoteProtocolError("\(field) must be 32 bytes") }
		return bytes
	}

	/// Constant-time comparison.
	public static func bytesEqual(_ a: Data, _ b: Data) -> Bool {
		guard a.count == b.count else { return false }
		var diff: UInt8 = 0
		for (left, right) in zip(a, b) { diff |= left ^ right }
		return diff == 0
	}

	private static func compare(_ a: Data, _ b: Data) -> Int {
		for (left, right) in zip(a, b) where left != right { return Int(left) - Int(right) }
		return a.count - b.count
	}
}

extension UInt32 {
	fileprivate func rotateLeft(_ count: UInt32) -> UInt32 { (self << count) | (self >> (32 - count)) }
}

public enum Base64URL {
	public static func encode(_ data: Data) -> String {
		data.base64EncodedString()
			.replacingOccurrences(of: "+", with: "-")
			.replacingOccurrences(of: "/", with: "_")
			.replacingOccurrences(of: "=", with: "")
	}

	public static func decode(_ text: String) throws -> Data {
		var normalized = text.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
		normalized += String(repeating: "=", count: (4 - normalized.count % 4) % 4)
		guard let data = Data(base64Encoded: normalized) else { throw RemoteProtocolError("value is not valid base64url") }
		return data
	}
}

File: apps/mobile/client-android/shared/src/commonMain/kotlin/org/vetta/android/domain/remote/protocol/RemoteCrypto.kt
package org.vetta.android.domain.remote.protocol

import java.security.MessageDigest
import java.security.SecureRandom
import org.bouncycastle.crypto.digests.SHA256Digest
import org.bouncycastle.crypto.generators.HKDFBytesGenerator
import org.bouncycastle.crypto.modes.ChaCha20Poly1305
import org.bouncycastle.crypto.params.AEADParameters
import org.bouncycastle.crypto.params.HKDFParameters
import org.bouncycastle.crypto.params.KeyParameter
import org.bouncycastle.crypto.params.X25519PrivateKeyParameters
import org.bouncycastle.crypto.params.X25519PublicKeyParameters
import org.bouncycastle.util.encoders.Base64

data class RemoteIdentityKeyPair(
    val secretKey: ByteArray,
    val publicKey: ByteArray,
) {
    override fun equals(other: Any?): Boolean =
        other is RemoteIdentityKeyPair &&
            secretKey.contentEquals(other.secretKey) &&
            publicKey.contentEquals(other.publicKey)

    override fun hashCode(): Int = 31 * secretKey.contentHashCode() + publicKey.contentHashCode()
}

data class RemoteSessionKeys(
    val sendKey: ByteArray,
    val receiveKey: ByteArray,
) {
    override fun equals(other: Any?): Boolean =
        other is RemoteSessionKeys && sendKey.contentEquals(other.sendKey) && receiveKey.contentEquals(other.receiveKey)

    override fun hashCode(): Int = 31 * sendKey.contentHashCode() + receiveKey.contentHashCode()
}

object RemoteCrypto {
    const val SEALED_ASSOCIATED_DATA = "vetta-remote-v2"
    private const val KEY_LENGTH = 32
    private const val NONCE_LENGTH = 24
    private const val INFO_PREFIX = "vetta-remote-v2"
    private val secureRandom = SecureRandom()

    fun randomBytes(length: Int): ByteArray = ByteArray(length).also(secureRandom::nextBytes)

    fun generateIdentityKeyPair(randomBytes: (Int) -> ByteArray = ::randomBytes): RemoteIdentityKeyPair =
        identityKeyPairFromSecret(randomBytes(KEY_LENGTH))

    fun identityKeyPairFromSecret(secretKey: ByteArray): RemoteIdentityKeyPair {
        if (secretKey.size != KEY_LENGTH) throw RemoteProtocolException("identity secret must be 32 bytes")
        val privateKey = X25519PrivateKeyParameters(secretKey, 0)
        return RemoteIdentityKeyPair(secretKey.copyOf(), privateKey.generatePublicKey().encoded)
    }

    fun deriveSessionKeys(
        role: RemoteRole,
        identity: RemoteIdentityKeyPair,
        ephemeral: RemoteIdentityKeyPair,
        peerIdentityKey: ByteArray,
        peerEphemeralKey: ByteArray,
    ): RemoteSessionKeys {
        val ephemeralShared = sharedSecret(ephemeral.secretKey, peerEphemeralKey)
        val staticToPeerEphemeral = sharedSecret(identity.secretKey, peerEphemeralKey)
        val ephemeralToPeerStatic = sharedSecret(ephemeral.secretKey, peerIdentityKey)
        val mobileFirst = role == RemoteRole.Mobile
        val ikm = concat(
            ephemeralShared,
            if (mobileFirst) staticToPeerEphemeral else ephemeralToPeerStatic,
            if (mobileFirst) ephemeralToPeerStatic else staticToPeerEphemeral,
        )
        val salt =
            if (mobileFirst) {
                concat(identity.publicKey, peerIdentityKey, ephemeral.publicKey, peerEphemeralKey)
            } else {
                concat(peerIdentityKey, identity.publicKey, peerEphemeralKey, ephemeral.publicKey)
            }
        val material = hkdf(ikm, salt, "$INFO_PREFIX/session".encodeToByteArray(), KEY_LENGTH * 2)
        val mobileToDesktop = material.copyOfRange(0, KEY_LENGTH)
        val desktopToMobile = material.copyOfRange(KEY_LENGTH, KEY_LENGTH * 2)
        return if (mobileFirst) {
            RemoteSessionKeys(mobileToDesktop, desktopToMobile)
        } else {
            RemoteSessionKeys(desktopToMobile, mobileToDesktop)
        }
    }

    fun sealFrame(
        key: ByteArray,
        frame: RemoteSessionFrame,
        associatedData: String = SEALED_ASSOCIATED_DATA,
        randomBytes: (Int) -> ByteArray = ::randomBytes,
    ): RemoteSealed {
        val nonce = randomBytes(NONCE_LENGTH)
        val ciphertext = xchacha(true, key, nonce, RemoteProtocol.encodeSession(frame).encodeToByteArray(), associatedData)
        return RemoteSealed(toBase64Url(nonce), toBase64Url(ciphertext))
    }

    fun openFrame(
        key: ByteArray,
        sealed: RemoteSealed,
        associatedData: String = SEALED_ASSOCIATED_DATA,
    ): RemoteSessionFrame {
        val plaintext =
            try {
                xchacha(false, key, fromBase64Url(sealed.nonce), fromBase64Url(sealed.ciphertext), associatedData)
            } catch (error: Throwable) {
                throw RemoteProtocolException("sealed frame failed authentication", error)
            }
        return try {
            RemoteProtocol.decodeSession(plaintext.decodeToString())
        } catch (error: Throwable) {
            throw RemoteProtocolException("sealed frame is not valid session JSON", error)
        }
    }

    fun verificationCode(identityKeyA: ByteArray, identityKeyB: ByteArray): String {
        val ordered = if (compareBytes(identityKeyA, identityKeyB) <= 0) concat(identityKeyA, identityKeyB) else concat(identityKeyB, identityKeyA)
        val digest = hkdf(ordered, ByteArray(0), "$INFO_PREFIX/sas".encodeToByteArray(), 4)
        val value =
            ((digest[0].toLong() and 0xff) shl 24) or
                ((digest[1].toLong() and 0xff) shl 16) or
                ((digest[2].toLong() and 0xff) shl 8) or
                (digest[3].toLong() and 0xff)
        return (value % 1_000_000).toString().padStart(6, '0')
    }

    fun sha256Hex(value: String): String =
        MessageDigest.getInstance("SHA-256").digest(value.encodeToByteArray()).joinToString("") { byte ->
            (byte.toInt() and 0xff).toString(16).padStart(2, '0')
        }

    fun randomToken(bytes: Int = 32): String = toBase64Url(randomBytes(bytes))

    fun decodePublicKey(value: String, field: String = "public key"): ByteArray {
        val bytes = fromBase64Url(value)
        if (bytes.size != KEY_LENGTH) throw RemoteProtocolException("$field must be 32 bytes")
        return bytes
    }

    fun toBase64Url(bytes: ByteArray): String =
        Base64.toBase64String(bytes).replace('+', '-').replace('/', '_').trimEnd('=')

    fun fromBase64Url(value: String): ByteArray {
        if (value.isEmpty() || !value.matches(Regex("^[A-Za-z0-9_-]+$"))) {
            throw RemoteProtocolException("value is not valid base64url")
        }
        val normalized = value.replace('-', '+').replace('_', '/')
        val padded = normalized + "=".repeat((4 - normalized.length % 4) % 4)
        return try {
            Base64.decode(padded)
        } catch (error: Throwable) {
            throw RemoteProtocolException("value is not valid base64url", error)
        }
    }

    fun bytesEqual(left: ByteArray, right: ByteArray): Boolean {
        if (left.size != right.size) return false
        var diff = 0
        for (index in left.indices) diff = diff or (left[index].toInt() xor right[index].toInt())
        return diff == 0
    }

    internal fun hchacha20(key: ByteArray, nonce: ByteArray): ByteArray {
        require(key.size == KEY_LENGTH && nonce.size == 16)
        val state = IntArray(16)
        state[0] = 0x61707865
        state[1] = 0x3320646e
        state[2] = 0x79622d32
        state[3] = 0x6b206574
        for (index in 0 until 8) state[4 + index] = word(key, index * 4)
        for (index in 0 until 4) state[12 + index] = word(nonce, index * 4)

        fun quarter(a: Int, b: Int, c: Int, d: Int) {
            state[a] += state[b]
            state[d] = (state[d] xor state[a]).rotateLeft(16)
            state[c] += state[d]
            state[b] = (state[b] xor state[c]).rotateLeft(12)
            state[a] += state[b]
            state[d] = (state[d] xor state[a]).rotateLeft(8)
            state[c] += state[d]
            state[b] = (state[b] xor state[c]).rotateLeft(7)
        }
        repeat(10) {
            quarter(0, 4, 8, 12)
            quarter(1, 5, 9, 13)
            quarter(2, 6, 10, 14)
            quarter(3, 7, 11, 15)
            quarter(0, 5, 10, 15)
            quarter(1, 6, 11, 12)
            quarter(2, 7, 8, 13)
            quarter(3, 4, 9, 14)
        }
        val output = ByteArray(32)
        listOf(0, 1, 2, 3, 12, 13, 14, 15).forEachIndexed { index, stateIndex ->
            writeWord(output, index * 4, state[stateIndex])
        }
        return output
    }

    private fun xchacha(
        encrypt: Boolean,
        key: ByteArray,
        nonce: ByteArray,
        input: ByteArray,
        associatedData: String,
    ): ByteArray {
        if (key.size != KEY_LENGTH || nonce.size != NONCE_LENGTH) {
            throw RemoteProtocolException("invalid key or nonce length")
        }
        val subkey = hchacha20(key, nonce.copyOfRange(0, 16))
        val ietfNonce = ByteArray(12).also { nonce.copyOfRange(16, 24).copyInto(it, 4) }
        val cipher = ChaCha20Poly1305()
        cipher.init(encrypt, AEADParameters(KeyParameter(subkey), 128, ietfNonce, associatedData.encodeToByteArray()))
        val output = ByteArray(cipher.getOutputSize(input.size))
        val processed = cipher.processBytes(input, 0, input.size, output, 0)
        val total = processed + cipher.doFinal(output, processed)
        return output.copyOf(total)
    }

    private fun sharedSecret(secretKey: ByteArray, peerPublicKey: ByteArray): ByteArray {
        if (secretKey.size != KEY_LENGTH || peerPublicKey.size != KEY_LENGTH) {
            throw RemoteProtocolException("X25519 keys must be 32 bytes")
        }
        val output = ByteArray(KEY_LENGTH)
        X25519PrivateKeyParameters(secretKey, 0).generateSecret(X25519PublicKeyParameters(peerPublicKey, 0), output, 0)
        return output
    }

    private fun hkdf(ikm: ByteArray, salt: ByteArray, info: ByteArray, length: Int): ByteArray {
        val generator = HKDFBytesGenerator(SHA256Digest())
        generator.init(HKDFParameters(ikm, salt, info))
        return ByteArray(length).also { generator.generateBytes(it, 0, it.size) }
    }

    private fun concat(vararg parts: ByteArray): ByteArray {
        val output = ByteArray(parts.sumOf(ByteArray::size))
        var offset = 0
        for (part in parts) {
            part.copyInto(output, offset)
            offset += part.size
        }
        return output
    }

    private fun compareBytes(left: ByteArray, right: ByteArray): Int {
        for (index in 0 until minOf(left.size, right.size)) {
            val difference = (left[index].toInt() and 0xff) - (right[index].toInt() and 0xff)
            if (difference != 0) return difference
        }
        return left.size - right.size
    }

    private fun word(bytes: ByteArray, offset: Int): Int =
        (bytes[offset].toInt() and 0xff) or
            ((bytes[offset + 1].toInt() and 0xff) shl 8) or
            ((bytes[offset + 2].toInt() and 0xff) shl 16) or
            ((bytes[offset + 3].toInt() and 0xff) shl 24)

    private fun writeWord(output: ByteArray, offset: Int, value: Int) {
        output[offset] = value.toByte()
        output[offset + 1] = (value ushr 8).toByte()
        output[offset + 2] = (value ushr 16).toByte()
        output[offset + 3] = (value ushr 24).toByte()
    }
}
